mdsa/algorithms/nystrom.py

```diff
@@ -0,0 +1,405 @@
+
+from mdsa.algorithm import Algorithm
+
+"""
+===================== Nystrom =====================
+
+Approximates an MDS mapping / a (partial) PCoA solution of an
+(unknown) full distance matrix using a k x n seed distance matrix.
+
+Use if you have a very high number of objects or if each distance
+caculation is expensive. Speedup comes from two factors: 1. Not all
+distances are calculated but only k x n. 2. Eigendecomposition is only
+applied to a k x k matrix.
+
+Calculations done after Platt (2005). See
+http://research.microsoft.com/apps/pubs/?id=69185 :
+``This paper unifies the mathematical foundation of three
+multidimensional scaling algorithms: FastMap, MetricMap, and Landmark
+MDS (LMDS). All three algorithms are based on the Nystrom
+approximation of the eigenvectors and eigenvalues of a matrix. LMDS is
+applies the basic Nystrom approximation, while FastMap and MetricMap
+use generalizations of Nystrom, including deflation and using more
+points to establish an embedding. Empirical experiments on the Reuters
+and Corel Image Features data sets show that the basic Nystrom
+approximation outperforms these generalizations: LMDS is more accurate
+than FastMap and MetricMap with roughly the same computation and can
+become even more accurate if allowed to be slower.``
+
+
+Assume a full distance matrix D for N objects:
+
+        /  E  |  F \
+    D = |-----|----|
+        \ F.t |  G /
+
+
+The correspondong association matrix or centered inner-product
+matrix K is:
+
+        /  A  |  B \
+    K = |-----|----|
+        \ B.t |  C /
+
+where A and B are computed as follows
+
+    A_ij = - 0.50 * (E_ij^2 -
+                 1/m SUM_p E_pj^2 -
+                 1/m SUM_q E_iq^2 +
+                 1/m^2 SUM_q E_pq^2
+
+B is computed as in Landmark MDS, because it's simpler and works
+better according to Platt):
+
+    B_ij = - 0.50 * (F_ij^2 - 1/m SUM_q E_iq^2)
+
+
+In order to approximate an MDS mapping for full matrix you only need E
+and F from D as seed matrix. This will mimick the distances for m seed
+objects. E is of dimension m x m and F of m x (N-m)
+
+E and F are then used to approximate and MDS solution x for the full
+distance matrix:
+
+
+x_ij =  sqrt(g_j) * U_ij, if i<=m
+        and
+        SUM_p B_pi U_pj / sqrt(g_j)
+
+where U_ij is the i'th component of the jth eigenvector of A (see
+below) and g_j is the j'th eigenvalue of A.  The index j only runs
+from 1 to k in order to make a k dimensional embedding.
+"""
+class Nystrom(Algorithm):
+    def __init__(self):
+        super(Nystrom, self).__init__(algorithm_name='nystrom')
+
+    def run(self, distance_matrix, num_dimensions_out=10):
+        n_seeds = int(log(distance_matrix.shape[0], 2)) ** 2
+        if n_seeds <= num_dimensions_out:
+            n_seeds = num_dimensions_out + 1
+        elif n_seeds >= distance_matrix.shape[0]:
+            n_seeds = distance_matrix.shape[0] - 1
+
+        sample_distmtx = array(sample(distance_matrix, n_seeds))
```

Comment:
does this work? python's `sample` is not aware of the structure of `np.array` and so I'd be surprised if the result of this is square. Testing locally, I get an exception:

```python
09:27:47 > import numpy as np

09:27:49 > a = np.arange(100).reshape(10, 10)

09:27:58 > a
Out[3]:
array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
       [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
       [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
       [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
       [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
       [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

09:27:59 > from random import sample

09:28:03 > sample(a, 10)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-c87f0dfcd12a> in <module>()
----> 1 sample(a, 10)

/Users/daniel/miniconda3/envs/py35-base/lib/python3.5/random.py in sample(self, population, k)
    309             population = tuple(population)
    310         if not isinstance(population, _Sequence):
--> 311             raise TypeError("Population must be a sequence or set.  For dicts, use list(d).")
    312         randbelow = self._randbelow
    313         n = len(population)

TypeError: Population must be a sequence or set.  For dicts, use list(d).
```